### PR TITLE
2016nano appx package support

### PR DESCRIPTION
**Pull request by wrightp**
Signed-off-by: Patrick Wright <patrick@chef.io>

When 2016nano is configured or auto-detected only appx packages are returned. Otherwise only msi packages are provided for windows platforms.

New acceptance test to validate the install.ps1 script installs appx packages for nano is passing.

This change **can not** be merged with 2.x as this branch already includes the 3.x changes. Apps dependent on these changes will need to update to mixlib-install 3.x.

@schisamo @scotthain @mwrock

### CODE REVIEW

File: lib/mixlib/install/backend/base.rb [hunk 1]
```diff
@@ -189,6 +189,11 @@ def windows_artifact_fixup!(artifacts)
             end
           end
 
+          # We currently only support appx packages for 2016nano
+          artifacts.delete_if do |artifact|
+            artifact.url.end_with?(".appx") unless %w{2016nano}.include?(artifact.platform_version)
+          end
+
```
Comment (mwrock) [inline lib/mixlib/install/backend/base.rb]: do we also need to delete the MSI artifact for nano?

Comment (wrightp) [inline lib/mixlib/install/backend/base.rb]: I suppose we should if they are built. I added the inverse of this into another block.

Comment (mwrock) [inline lib/mixlib/install/backend/base.rb]: ah cool

File: lib/mixlib/install/backend/base.rb [hunk 2]
```diff
@@ -189,6 +189,15 @@ def windows_artifact_fixup!(artifacts)
             end
           end
 
+          # We currently only support appx packages for 2016nano
+          artifacts.delete_if do |artifact|
+            artifact.url.end_with?(".appx") unless %w{2016nano}.include?(artifact.platform_version)
```
Comment (schisamo) [inline lib/mixlib/install/backend/base.rb]: Let's move the list of nano platform versions up in a `SUPPORTED_WINDOWS_NANO_VERSIONS` constant like we have for `SUPPORTED_WINDOWS_DESKTOP_VERSIONS`.

**Comment by wrightp** [comment]
@schisamo I'll test where you noted. I also mistakenly thought the 2016 -> 2012r logic that already was being called for desktop versions. (which it isn't)

File: lib/mixlib/install/options.rb [hunk 3]
```diff
@@ -213,8 +218,15 @@ def validate_platform_options
         end
       end
 
-      def map_windows_desktop_versions!
-        options[:platform_version] = Util.map_windows_desktop_version(platform_version)
+      def map_windows_versions!
+        return unless for_windows?
+
+        # Store original platform version for nano requests
+        if SUPPORTED_WINDOWS_NANO_VERSIONS.include?(options[:platform_version])
```
Comment (schisamo) [inline lib/mixlib/install/options.rb]: Feels like we should always set `@original_platform_version`, I would remove this guard check.

File: lib/mixlib/install/backend/base.rb [hunk 4]
```diff
@@ -123,14 +121,21 @@ def filter_artifacts(artifacts)
             return closest_compatible_artifact
           end
 
-           # Return an exception if we get to the end of the method
+          # Determine which platform version should be displayed in the exception
+          platform_version = if options.original_platform_version.nil?
+                               options.platform_version
+                             else
+                               options.original_platform_version
+                             end
```
Comment (schisamo) [inline lib/mixlib/install/backend/base.rb]: Related to https://github.com/chef/mixlib-install/pull/191/files#r106671866 we can skip this code if we just ensure `options.original_platform_version` is always set....we could then just do this below:

```ruby
platform version: #{options.original_platform_version}
```

Comment (wrightp) [inline lib/mixlib/install/backend/base.rb]: totally was thinking that!

File: lib/mixlib/install/options.rb (hunk 3 above)
Comment (wrightp) [inline lib/mixlib/install/options.rb]: aye

File: .travis.yml [hunk 5]
```diff
@@ -14,10 +14,13 @@ addons:
...
   include:
-  - rvm: 2.2.5
-    env: CHEF_ACCEPTANCE=false
-  - rvm: 2.3.0
-    env: CHEF_ACCEPTANCE=true
+  - rvm: 2.2
+  - rvm: 2.3
+    env: SUITE=ubuntu
+  - rvm: 2.3
+    env: SUITE=powershell
```
Comment (schisamo) [inline .travis.yml]: Can we call this suite something like `windows` and the new one `windows-nano`?

File: acceptance/2016nano/.acceptance/acceptance-cookbook/recipes/provision.rb [hunk 6]
```diff
@@ -0,0 +1,41 @@
+run_tf "plan"
+
+run_tf "apply"
+
+ruby_block "install via winrm api" do
+  block do
+    # We do this hackery to ensure we load mixlib-install from source
+    install_command = "install -project angry-omnibus-toolchain -channel current -version 1.1.57"
```
Comment (schisamo) [inline acceptance/2016nano/.acceptance/acceptance-cookbook/recipes/provision.rb]: Just a heads up `angry-omnibus-toolchain 1.1.57` will eventually be reaped from the `current` channel. We should land a release in the `stable` channel (which is never reaped) before that is an issue.

File: acceptance/2016nano/inspec/verify.rb [hunk 7]
```diff
@@ -0,0 +1,3 @@
+describe file("c:/opscode/angry-omnibus-toolchain") do
+  it { should be_directory }
+end
```
Comment (schisamo) [inline acceptance/2016nano/inspec/verify.rb]: Can we add some checks that use [InSpec's command resource](http://inspec.io/docs/reference/resources/command/) to ensure a few of the toolchain's actual tools are installed and are functional. Maybe check for `git`, `ruby` and `gcc`? 

If the`PATH` is configured per:
https://github.com/chef-cookbooks/omnibus/blob/master/recipes/_environment.rb#L24-L28

You can easily just call `--version` on each of those tools...like:
https://github.com/chef-cookbooks/omnibus/blob/master/recipes/_environment.rb#L57-L61

File: acceptance/2016nano/terraform/application.tf [hunk 8]
```diff
@@ -0,0 +1,59 @@
+data "aws_ami" "windows_ami" {
```
Comment (schisamo) [inline acceptance/2016nano/terraform/application.tf]: `windows_server_nano_ami`

File: acceptance/2016nano/terraform/application.tf [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+    values = ["gp2"]
+  }
+
+  filter {
+    name   = "image-type"
+    values = ["machine"]
+  }
+}
+
+resource "aws_instance" "mixlib_install_ps1" {
```
Comment (schisamo) [inline acceptance/2016nano/terraform/application.tf]: This Terraform resource should be named something different than `mixlib_install_ps1 `...right?

File: acceptance/2016nano/terraform/application.tf [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+  vpc_security_group_ids = [
+    "sg-96274af3",
+  ]
+
+  tags {
+    # ChefOps's AWS standard tags:
+    X-Dept        = "EngServ"
+    X-Contact     = "pwright"
+    X-Production  = "false"
+    X-Environment = "development"
```
Comment (schisamo) [inline acceptance/2016nano/terraform/application.tf]: ` X-Environment = "acceptance"`

File: lib/mixlib/install/backend/base.rb [hunk 11]
```diff
@@ -146,6 +144,15 @@ def windows_artifact_fixup!(artifacts)
           new_artifacts = [ ]
           native_artifacts = [ ]
 
+          # We only return appx packages when a nano platform version is requested.
+          if options.class::SUPPORTED_WINDOWS_NANO_VERSIONS.include?(options.original_platform_version)
+            return artifacts.find_all { |a| appx_artifact?(a) }
```
Comment (schisamo) [inline lib/mixlib/install/backend/base.rb]: Let's just add a `#appx_artifact?` method to `Mixlib::Install::ArtifactInfo`.

File: .travis.yml (hunk 5 above)
Comment (wrightp) [inline .travis.yml]: sure, I'll rename the directory and update this.

File: acceptance/2016nano/.acceptance/acceptance-cookbook/recipes/provision.rb (hunk 6 above)
Comment (wrightp) [inline acceptance/2016nano/.acceptance/acceptance-cookbook/recipes/provision.rb]: sure when that happens I'll update the test

File: acceptance/2016nano/inspec/verify.rb (hunk 7 above)
Comment (wrightp) [inline acceptance/2016nano/inspec/verify.rb]: absolultely

File: acceptance/2016nano/terraform/application.tf (hunk 9 above)
Comment (wrightp) [inline acceptance/2016nano/terraform/application.tf]: these names could be foobar. doesn't really matter. it was a copy/paste remnant. I'll update it though.

File: acceptance/2016nano/terraform/application.tf (hunk 10 above)
Comment (wrightp) [inline acceptance/2016nano/terraform/application.tf]: I thought `acceptance` was reserved for Delivery acceptance. No?

File: lib/mixlib/install/backend/base.rb (hunk 11 above)
Comment (wrightp) [inline lib/mixlib/install/backend/base.rb]: done

File: acceptance/2016nano/terraform/application.tf (hunk 10 above)
Comment (schisamo) [inline acceptance/2016nano/terraform/application.tf]: @wrightp Nope....it represents the application specified in `X-Application`.

Comment (wrightp) [inline acceptance/2016nano/terraform/application.tf]: ok I'll update the other suites as well

**Comment by wrightp** [comment]
@schisamo I believe I accounted for all your comments. Please review.

**Comment by schisamo** [comment]
This is so awesome for so many reasons...thank you for struggling through how to acceptance test Windows Server Nano using Terraform and InSpec!

**Pull request merged**